dev-tools/omdb/src/bin/omdb/mgs/dashboard.rs

```diff
@@ -0,0 +1,1115 @@
+    }
+
+    Ok(false)
+}
+
+fn secs() -> Result<u64> {
+    let now = SystemTime::now().duration_since(SystemTime::UNIX_EPOCH)?;
+    Ok(now.as_secs())
+}
+
+///
+/// Runs `omdb mgs dashboard`
+///
+pub(crate) async fn cmd_mgs_dashboard(
+    omdb: &crate::Omdb,
+    log: &slog::Logger,
+    mgs_args: &crate::mgs::MgsArgs,
+    args: &DashboardArgs,
+) -> Result<(), anyhow::Error> {
+    let mut input = if let Some(ref input) = args.sensors_args.input {
+        let file = File::open(input)?;
+        SensorInput::CsvReader(
+            csv::Reader::from_reader(file),
+            csv::Position::new(),
+        )
+    } else {
+        SensorInput::MgsClient(mgs_args.mgs_client(omdb, log).await?)
+    };
+
+    let (metadata, values) =
+        sensor_metadata(&mut input, &args.sensors_args).await?;
+
+    //
+    // A bit of shenangians to force metadata to be 'static -- which allows
+    // us to share it with tasks.
+    //
+    let metadata = Box::leak(Box::new(metadata));
+    let metadata: &_ = metadata;
+
+    let mut dashboard = Dashboard::new(metadata)?;
+    let mut last = values.time;
+    let mut force = true;
+    let mut update = true;
+
+    dashboard.values(&values);
+
+    if args.sensors_args.input.is_some() && !args.simulate_realtime {
+        loop {
+            let values = sensor_data(&mut input, metadata).await?;
+
+            if values.time == 0 {
+                break;
+            }
+
+            if values.time != last + 1 {
+                dashboard.gap(values.time - last - 1);
+            }
+
+            last = values.time;
+            dashboard.values(&values);
+        }
+
+        update = false;
+    }
+
+    // setup terminal
+    enable_raw_mode()?;
+    let mut stdout = io::stdout();
+    execute!(stdout, EnterAlternateScreen, EnableMouseCapture)?;
+    let backend = CrosstermBackend::new(stdout);
+    let mut terminal = Terminal::new(backend)?;
+
+    let res = 'outer: loop {
+        match run_dashboard(&mut terminal, &mut dashboard, force) {
+            Err(err) => {
+                break Err(err);
+            }
+            Ok(true) => {
+                break Ok(());
+            }
+            _ => {}
+        }
+
+        force = false;
+        let now = secs()?;
+
+        if update && now != last {
+            let kicked = Instant::now();
+            let f = sensor_data(&mut input, metadata);
+            last = now;
+
+            while Instant::now().duration_since(kicked).as_millis() < 800 {
+                tokio::time::sleep(Duration::from_millis(10)).await;
+
+                match run_dashboard(&mut terminal, &mut dashboard, force) {
+                    Err(err) => {
+                        break 'outer Err(err);
+                    }
+                    Ok(true) => {
+                        break 'outer Ok(());
+                    }
+                    _ => {}
+                }
+            }
+
+            let values = f.await?;
```

Comment:
Looking at this again while I was poking around with the terminal restore on error question: why do we construct `f` before the while loop but not `.await` it until here? I think it would be equivalent to get rid of `f` and call `sensor_data(...).await` directly here, but maybe there was a different intent with `f`?